#  第十一章：并发编程实践

## 目录

- [并发组件ConcurrentHashMap使用注意事项](#1)
- [SImpleDateFormat是线程不安全d额](#2)
  - [问题复现](#3)
  - [问题分析](#4)
- [使用Timer时需要注意的问题](#5)
  - [问题复现](#6)
  - [实现原理分析](#7)
  - [小结](#8)
- [创建线程和线程池时要指定与业务相关的名称](#9)
  - [创建线程需要有线程名](#10)
  - [创建线程时也需要指定线程的名字](#11)
  - [小结](#12)
- [使用线程池的情况下当程序结束时记得调用shutdown关闭线程池](#13)
  - [问题复现](#14)
  - [问题分析 ](#15)
  - [小结](#16)
- [使用ThreadLocal不当可能会导致内存泄漏](#17)
  - [为何会出现内存泄漏](#18)
  - [在线程池中使用ThreadLocal导致内存泄漏](#19)

  

<h2 id='1'>并发组件ConcurrentHashMap使用注意事项</h2>
ConcurrentHashMap 虽然为并发安全的组件，但是使用不当仍然会导致程序错误。 本节通过简单的案例来复现这些问题，并给出开发时如何避免的策略。这里借用直播的一个场景，在直播业务中，每个直播间对应一个 topic，每个用户进入直播间H才会把自己设备的ID绑定到这个topic上， 也就是一个topic对应一堆用户设备。
可以使用 map 来维护这些信息，其中 key 为 topic, value 为设备的 list。下面使用代码来模拟多用户同时进入直播间时map信息的维护。注意JSON调用的是阿里写的FastJSON包，可在github上下载jar包。然后导入IDEA

```java
public class TestMap2 {
    static ConcurrentHashMap<String, List<String>> map =new ConcurrentHashMap<String, List<String>>();
    public static void main(String[]args){
        //(2)
        Thread threadOne = new Thread(new Runnable() {
            @Override
            public void run() {
                List<String> list1 = new ArrayList<>();
                list1.add("device1");
                list1.add("device2");
                List<String> oldList = map.putIfAbsent("topic1",list1);
                if(null != oldList){
                    oldList.addAll(list1);
                }
                System.out.println(JSON.toJSONString(map));
            }

        });
        Thread threadTwo = new Thread(new Runnable() {
            @Override
            public void run() {
                List<String> list1 = new ArrayList<>();
                list1.add("device11");
                list1.add("device22");
                List<String> oldList = map.putIfAbsent("topic1",list1);
                if(null != oldList){
                    oldList.addAll(list1);
                }
                System.out.println(JSON.toJSONString(map));
            }

        });
        Thread threadThree = new Thread(new Runnable() {
            @Override
            public void run() {
                List<String> list1 = new ArrayList<>();
                list1.add("device111");
                list1.add("device222");
                List<String> oldList = map.putIfAbsent("topic2",list1);
                if(null != oldList){
                    oldList.addAll(list1);
                }
                System.out.println(JSON.toJSONString(map));
            }

        });
        threadOne.start();
        threadTwo.start();
        threadThree.start();
    }
}

```

运行结果如下

```java
{"topic1":["device11","device22"],"topic2":["device11","device22"]}
{"topic1":["device11","device22"],"topic2":["device11","device22"]}
{"topic1":["device11","device22"],"topic2":["device11","device22"]}

```

可见，topic1房间的用户丢失了部分，这是因为put方法如果发现map里面存在这个key，则使用value覆盖该key对应的老的value值。而putIfAbsent方法则是，如果发现已经存在该key则返回该key对应的value，单步进行覆盖，如果不存在则新增该key，并且判断和写入的是原子性操作。使用putIfAbsent替代后的代码如下:

```java
public class TestMap2 {
    static ConcurrentHashMap<String, List<String>> map =new ConcurrentHashMap<String, List<String>>();
    public static void main(String[]args){
        //(2)
        Thread threadOne = new Thread(new Runnable() {
            @Override
            public void run() {
                List<String> list1 = new ArrayList<>();
                list1.add("device1");
                list1.add("device2");
                List<String> oldList = map.putIfAbsent("topic1",list1);
                if(null != oldList){
                    oldList.addAll(list1);
                }
                System.out.println(JSON.toJSONString(map));
            }

        });
        Thread threadTwo = new Thread(new Runnable() {
            @Override
            public void run() {
                List<String> list1 = new ArrayList<>();
                list1.add("device11");
                list1.add("device22");
                List<String> oldList = map.putIfAbsent("topic1",list1);
                if(null != oldList){
                    oldList.addAll(list1);
                }
                System.out.println(JSON.toJSONString(map));
            }

        });
        Thread threadThree = new Thread(new Runnable() {
            @Override
            public void run() {
                List<String> list1 = new ArrayList<>();
                list1.add("device111");
                list1.add("device222");
                List<String> oldList = map.putIfAbsent("topic2",list1);
                if(null != oldList){
                    oldList.addAll(list1);
                }
                System.out.println(JSON.toJSONString(map));
            }

        });
        threadOne.start();
        threadTwo.start();
        threadThree.start();
    }
}

```

<h2 id='2'>SimpleDateFormat是线程不安全的</h2>
SimpleDateFom1at 是 Java 提供的 一个格式化和解析日期的工具类，在日常开发中经常会用到，但是由于它是线程不安全的，所以多线程共用一个 SimpleDateFormat实例对日期进行解析或者格式化会导致程序出错。本节来揭示它为何是线程不安全的，以及如何避免该问题 。

<h3 id='3'>问题复现</h3>
```java
public class TestDateFormat {

    static SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");

    public static void main(String []args) {
        for(int i=0;i<10;i++){
            Thread thread = new Thread(new Runnable() {
                @Override
                public void run() {
                    try {
                        System.out.println(sdf.parse("2019-11-05 09:21:25"));
                    }catch (ParseException e){
                        e.printStackTrace();
                    }
                }
            });
            thread.start();
        }
    }
   }
```

多运行几次代码就会抛出 java.lang.NumberFormatException 异常， 增加线程的个数有利于复现该问题 。

<h3 id='4'>问题分析</h3>
![image-20191115211354952](/Users/hzx/Library/Application Support/typora-user-images/image-20191115211354952.png)

由类图结构可以看到每个SimpleDateFormat实例里面都有一个Calendar对象，后面我们就会知，SimpleDateFormat之所以是线程不安全的，就是因为Calendar是线程不安全的。后者由于其中存放的日期数据的变量都是线程不安全的，比如fields，time等。

下面从代码层面来看parse方法做了什么事情

```java
public Date parse(String text, ParsePosition pos){
		//(1)解析日期字符串，并将解析好的数据放入CalendarBuilder的实例calb中
		...
		Date parseDate;
		try{//(2)使用calb中解析好的日期数据设置calendar
				parseDate = calb.establish(calendar).getTime();
				...
		}catch(IllegalArgumentException e){
				...
				return null;
		}
		return parseDate;
}
```

代码（1）的作用主要是解析日期字符串并把解析好的数据放入CalendarBuilder的实例calb中。CanlendarBuilder是一个建造者模式，用来存放后面需要的数据

代码（2）使用calb中解析好的日期数据设置calendar，calb.establish的代码如下。

```
Calendar establish(Calendar cal){
	...
	//(3)重置日期对象cal的属性值
	cal.clear();
	(4)使用calb中的属性设置cal
	...
	(5)返回设置好的cal对象
	return cal;
}
```

代码(3)重置Calendar对象里面的属性值，如下所示

```java
public final void clear(){
	for(int i=0;i<fields.length){
    stamp[i] = fields[i] = 0;//UNSET == 0 
    isSet[i++] = false;
  }
  areAllFieldsSet = are FieldsSet = false;
  isTimeSet = false;
}
```

从上面代码看出，代码（3）（4）（5）不是原子性操作。当多个线程调用了parse方法时，比如线程A执行了代码（3）和代码（4），也就是设置好了cal对象，但是在执行代码（5）之前，线程B执行了代码（3），清空了cal对象。由于多个线程使用的是一个cal对象，所以线程A执行代码（5）返回的可能就是被线程B清空的对象，当然也有可能线程B执行了代码（4），设置被线程A修改的cal对象，从而导致程序出现错误。

解决办法有三种

- 每个使用时读new一个SimpleDateFormat的实例，这样可以保证每个实例使用自己的Calendar实例，但是每次使用都需要new一个对象，并且使用后由于没有其他引用，又需要回收，开销很大
- 让代码（3）（4）（5）成为原子性操作，可以使用synchronized进行同步。这样进行同步意味着多个线程要竞争锁，在高并发场景下这会导致系统响应性能下降。
- 使用ThreadLocal，这样每个线程都只需要使用一个SimpleDateFormat实例，这相比第一种方式大大节省了对象的创建销毁开销，并且不需要使多个线程同步。使用代码如下

```java

public class TestSimpleDateFormat2 {

    static ThreadLocal<DataFormat> safeSdf = new ThreadLocal<DataFormat>(){
        @Override
        protected SimpleDateFormat initialValue(){
            return new SimpleDateFormat();
        }
    };

    public  static  void main(String[]args){
        for(int i=0;i<10;i++){
            Thread thread = new Thread(new Runnable() {
                @Override
                public void run() {
                    try{
                        System.out.println(safeSdf.get().parse("2017-12-13 15:17:27"));
                    }catch (ParseException e){
                        e.printStackTrace();
                    }finally {
                        safeSdf.remove();
                    }
                }
            });
            thread.start();
        }
    }


}
```

尴尬，上面的代码有bug。。。

<h2 id='5'>使用Timer时需要注意的问题</h2>
当一个Timer运行多个TimerTask时，只要其中一个TimerTask在执行中向run方法外抛出了异常，其他任务也会自动终止。

<h3 id='6'>问题复现</h3>
```java

public class TestTimer {
    static Timer timer = new Timer();

    public static void main(String[]args){
        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                System.out.println("one task");
                try{
                    Thread.sleep(1000);

                }catch(Exception e){
                    e.printStackTrace();
                }
                throw new RuntimeException("error");
            }
        },500);

        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                for (; ; ) {
                    System.out.println("two task");
                    try {
                        Thread.sleep(1000);

                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                    throw new RuntimeException("error");
                }
            }
        },1000);
    }
}
```

如上代码首先添加了第 一 个任 务 ， 让其在 500ms 后执行。然后添加了第二个任务在1s后执行，我们期望当第一个任务输出一oneTask---后，等待1s，第二个任务输出一two Task---，但是执行代码后，输出结果为 

```java
one task
Exception in thread "Timer-0" java.lang.RuntimeException: error
	at com.hzx.hzx20191115.TestTimer$1.run(TestTimer.java:20)
	at java.util.TimerThread.mainLoop(Timer.java:555)
	at java.util.TimerThread.run(Timer.java:505)
```

<h3 id='7'>Timer实现原理分析</h3>
![image-20191115230837597](/Users/hzx/Library/Application Support/typora-user-images/image-20191115230837597.png)

- TaskQueue是一个由平衡二叉树堆实现的优先级队列，每个 Timer对象内部有一个 TaskQueue 队列。 用户线程调用 Timer 的schedule 方法就是把 TimerTask 任务添加到 TaskQueue 队列。在调用 schedule 方法时 ， long delay参数用来指明该任务延迟多少时间执行。 

- TimerThread是具体执行任务的线程，它从TaskQueue 队列里面获取优先级最高的任务进行执行。需要注意 的是，只有执行完了当前的任务才会从队列里获取下一个任务 ，而不管队列里是否有任务已经到了设置的 delay 时间。一个 Timer 只有一个 TimerThread 线程，所以可知 Timer 的内部实现是一个多生产者 - 单消费者模型。 

  从该实现模型我们知道，要探究上面的问题只需研究 TimerThread 的实现就可以了 。 

TimerThread的run方法的主要逻辑代码如下

```java
public void run (){
	try{
    mainLoop();
  }finally{
    //Someone killed this Thread, behave as if Timer cancelled 
    synchronized(queue){
      newTaskMayBeScheduled = false;
      queue.clear();  //Eliminate obsolete reference
    }
  }
}

private void mainLoop(){
  while(true ){
    try{
      TimerTask task;
      boolean taskFired;
      //从队列里面获取任务时要加锁
      synchronized(queue){
        ...
      }
      if(taskFired) task.run();//执行任务
    }catch(InterruptedException e) {
      
    }
  }
}
```

当任务在执行过程中抛出InterruptedException之外的异常时，唯一的消费线程就会因为抛出异常而终止，那么队列里面的其他待执行任务就会被清除。所以在TimerTask的run方法内最好使用try-catch结构捕获可能的异常，不要让异常抛到run方法之外。其实要实现 Timer 功能，使用 ScheduledThreadPoo!Executor 的 schedule 是比较好的选择。如果ScheduledThreadPoo!Executor 中的一个任务抛出异常，其他任务则不受影响。

```java
public class TestScheduledThreadPoolExecutor {
    static ScheduledThreadPoolExecutor scheduledThreadPoolExecutor = new ScheduledThreadPoolExecutor(1);
    public static void main(String[]args){
        scheduledThreadPoolExecutor.schedule(new Runnable() {
            @Override
            public void run() {
                System.out.println("--one Task");
                try{
                    Thread.sleep(1000);
                }catch (InterruptedException e){
                    e.printStackTrace();
                }
                throw new RuntimeException("error");
            }
        },500, TimeUnit.MICROSECONDS);

        scheduledThreadPoolExecutor.schedule(new Runnable() {
            @Override
            public void run() {
                for (int i = 0; i < 2; i++) {
                    System.out.println("--two Task");
                    try {
                        Thread.sleep(1000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        },1000, TimeUnit.MICROSECONDS);
        scheduledThreadPoolExecutor.shutdown();
    }
}
```

结果如下

```java
--one Task
--two Task
--two Task
```

<h3 id='8'>小结</h3>
ScheduledThreadPoolExecutor 是并发包提供的组件，其提供的功能包含但不限于TimeroTimer 是固定的多线程生产单线程消费，但是 ScheduledThreadPoolExecutor 是可以配置的，既可以是多线程生产单线程消费 也可以是多线程生产多线程消费，所以在日常开发中使用定时器功能时应该优先使用ScheduledThreadPoolExecutor。

<h2 id='9'>创建线程和线程池时要指定与业务相关的名称</h2>
在日常开发中，当一个应用需要创建多个线程或者线程池最好给每个线程或者线程池根据业务类型设置具体的名称，以便在出现问题时方便进行定位。下面就通过实例来说明不设置违和难以定位问题，以及如何进行设置。

<h3 id='10'>创建线程需要有线程名</h3>
下面通过简单的代码来说明不指定线程名称为何难定位问题，代码如下。

```java
public class Test {
    public static void main(String[]args){
        Thread threadOne = new Thread(new Runnable() {
            @Override
            public void run() {
                System.out.println("保存订单的线程 ");
                try{
                    Thread.sleep(500);
                }catch (InterruptedException e){
                    e.printStackTrace();
                }
                throw new NullPointerException();
            }
        });

        Thread threadTwo = new Thread(new Runnable() {
            @Override
            public void run() {
                System.out.println("保存收获地址的线程");
            }
        });
        threadOne.start();
        threadTwo.start();
    }
}
```

输入如下

```java
保存收获地址的线程
保存订单的线程 
Exception in thread "Thread-0" java.lang.NullPointerException
	at com.hzx.hzx20191115.Test$1.run(Test.java:14)
	at java.lang.Thread.run(Thread.java:748)
```

从运行结果上看，Thread-0抛出了NPE异常，单看日志根本无法判断你是订单模块的线程抛出的异常。首先我们分析下Thread-0是怎么来的。看下创建线程的代码。

```java
public Thread(Runnalbe target){
	init (null, target, "Thread-"+nextThreadNum(),0);
}
private void init(ThreadGroup g, Runnable target, String name, long stackSize){
	init(g, target, name, stackSize null);
}
```

由这段代码可知，如果调用没有指定线程名称的方法创建线程，其内部会使用"Thead-"+nextThreadNum()作为线程的默认名称，其中nextThreadNum的代码如下。

```java
private static int threadInitNumber
private static synchronized int nextThreadNum(){
	return threadInitNumber++;
}
```

由此可知， threadlnitNumber是 static变量，nextThreadNum是 static方法，所以线程的编号是全应用唯一的并且是递增的。 这里由于涉及多线程递增 threadlnitNumber,也就是执行读取-递增一写入操作 ，而这 是线程不安全的，所以要使用方法级别的synchronized 进行同步 。

当一个系统中有多个业务模块而每个模块又都使用自己的线程时，除非抛出与业务相关的异常，否则你根本没法判断是哪一个模块出现了问题 。现在修改代码如下 。

```java
public class Test2 {
    static final String THREAD_SAVE_ORDER = "THREAD_SAVE_ORDER";
    static final String THREAD_SAVE_ADDR = "THREAD_SAVE_ADDR";

    public static void main(String[]args){
        //订单模块
        Thread threadOne = new Thread(new Runnable() {
            @Override
            public void run() {
                System.out.println("保存订单的线程");
                throw new NullPointerException();
            }
        },THREAD_SAVE_ORDER);

        Thread threadTwo = new Thread(new Runnable() {
            @Override
            public void run() {
                System.out.println("保存订单的线程");
                throw new NullPointerException();
            }
        },THREAD_SAVE_ADDR);
        threadOne.start();
        threadTwo.start();
    }



}
```

上面的代码在创建线程给线程指定了一个与具体业务模块相关的名称，运行代码，输出结果为

```java
保存订单的线程
保存收货地址的线程
Exception in thread "THREAD_SAVE_ORDER" java.lang.NullPointerException
	at com.hzx.hzx20191115.Test2$1.run(Test2.java:13)
	at java.lang.Thread.run(Thread.java:748)
```

从运行结果就可以定位到是保存订单模块抛出了NPE异常。

<h3 id='11'>
  创建线程池时也需要指定线程池的名称
</h3>

同理下面通过简单的代码来说明不指定线程池名称为何难定位问题

```java
public class TestThreadPool {
    static ThreadPoolExecutor executorone = new ThreadPoolExecutor(5,5,1, TimeUnit.MINUTES, new LinkedBlockingDeque<>());
    static ThreadPoolExecutor executorTwo = new ThreadPoolExecutor(5,5,1, TimeUnit.MINUTES, new LinkedBlockingDeque<>());

    public static void main(String[]args){
        //接受用户连接模块
        executorone.execute(new Runnable() {
            @Override
            public void run() {
                System.out.println("接受用户连接线程");
                throw new NullPointerException();
            }
        });
        //具体处理用户模块
        executorTwo.execute(new Runnable() {
            @Override
            public void run() {
                System.out.println("具体处理业务请求线程");
            }
        });
        executorone.shutdown();
        executorTwo.shutdown();
    }
}

```
结果如下
```java
接受用户连接线程
Exception in thread "pool-1-thread-1" java.lang.NullPointerException
	at com.hzx.hzx20191115.TestThreadPool$1.run(TestThreadPool.java:17)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
具体处理业务请求线程
```

同样我们并不知道哪个模块的线程池抛出了异常，那么我们看下pool-1-thread-1如何来。这里使用了线程池默认的ThreadFactory，查看线程池创建的源码如下

```java
  public ThreadPoolExecutor(int corePoolSize,
                              int maximumPoolSize,
                              long keepAliveTime,
                              TimeUnit unit,
                              BlockingQueue<Runnable> workQueue) {
        this(corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue,
             Executors.defaultThreadFactory(), defaultHandler);
    }
    
    public static ThreadFactory defaultThreadFactory(){
    	return new DefaultThreadFactory();
    }

static class DefaultThreadFactory implements ThreadFactory {
  			//(1)
        private static final AtomicInteger poolNumber = new AtomicInteger(1);
        private final ThreadGroup group;
  			//(2)
        private final AtomicInteger threadNumber = new AtomicInteger(1);
  			//(3)
        private final String namePrefix;

        DefaultThreadFactory() {
            SecurityManager s = System.getSecurityManager();
            group = (s != null) ? s.getThreadGroup() :
                                  Thread.currentThread().getThreadGroup();
            namePrefix = "pool-" +
                          poolNumber.getAndIncrement() +
                         "-thread-";
        }

        public Thread newThread(Runnable r) {
          //(4)
            Thread t = new Thread(group, r,
                                  namePrefix + threadNumber.getAndIncrement(),
                                  0);
            if (t.isDaemon())
                t.setDaemon(false);
            if (t.getPriority() != Thread.NORM_PRIORITY)
                t.setPriority(Thread.NORM_PRIORITY);
            return t;
        }
    }
```

代码（1）中的 poo!Number是 static 的原子变量，用来记录当前线程池的编号，它是 应用级别的，所有线程池共用一个，比如创建第一个线程池时线程池编号为1，创建第二个线程池时线程池的编号为 2，所以 pool-1-thread-1 里面的 pool-1 中的1就是这个值。 

代码（2）中的threadNumber是线程池级别的，每个线程池使用该变量来记录该线程池中线程的编号 ， 所以 pool-1-thread-1里面的thread-I 中的1就是这个值。 

代码（3）中的 namePrefix 是线程池中线程名称的前缀，默认固定为 pool 

代码(4)具体创建线程，线程的名称是使用 namePrefix+threadNumber. getAndlncrement()拼接的。
 由此我们知道，只需对 DefaultThreadFactory 的代码中的 namePrefix 的初始化做下手脚， 

即当需要创建线程池时传入与业务相关的 namePrefix名称就可以了，代码如下。 

```java
public class NamedThreadFactory {

    static class NamedThreadFacotry implements ThreadFactory{
        private static final AtomicInteger poolNumber = new AtomicInteger(1);
        private final ThreadGroup group;
        private final AtomicInteger threadNumber = new AtomicInteger(1);
        private final String namePrefix;
        NamedThreadFacotry(String name){
            SecurityManager s= System.getSecurityManager();
            group = (s!=null)? s.getThreadGroup(): Thread.currentThread().getThreadGroup();
            if(null == name || name.isEmpty()){
                name ="pool";
            }
            namePrefix = name+"-"+poolNumber.getAndIncrement()+"-thread-";
        }
        @Override
        public Thread newThread(Runnable r) {
            Thread t =new Thread(group, r,namePrefix +threadNumber.getAndIncrement(),0);
            if(t.isDaemon()){
                t.setDaemon(false);
            }
            if(t.getPriority()!= Thread.NORM_PRIORITY){
                t.setPriority(Thread.NORM_PRIORITY);
            }
            return t;
        }
    }
```

重新之前的测试程序，创建线程池如下

```java
    static ThreadPoolExecutor  executorOne = new ThreadPoolExecutor(5,5,1, TimeUnit.MINUTES,new LinkedBlockingDeque<>(),new NamedThreadFactory.NamedThreadFacotry("ASYN-Accpet-POOL"));

    static ThreadPoolExecutor  executorTwo = new ThreadPoolExecutor(5,5,1, TimeUnit.MINUTES,new LinkedBlockingDeque<>(),new NamedThreadFactory.NamedThreadFacotry("ASYN-PROCESS-POOL"));

```

结果如下

```java
接受用户连接线程
Exception in thread "ASYN-Accpet-POOL-1-thread-1" java.lang.NullPointerException
	at com.hzx.hzx20191115.TestThreadPool$1.run(TestThreadPool.java:17)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
具体处理业务请求线程
```

<h3 id='12'>小结</h3>
本节通过简单的例子介绍了为何不为线程或者线程池起名字会给问题排查带来麻烦，然后通过源码分析介绍了线程和线程池名称及默认名称是如何来的，以及如何定义线程池名称以便追溯问题。 另外，在 run方法内使用 try-catch块，避免将异常抛到 run方法之外，同时打印日志也是一个最佳实践 。



<h2 id='13'>使用线程池的情况下当程序结束时记得调用
shutdown 关闭线程池</h2>
如果不通过调用线程池对象的shutdown方法关闭线程池，则当线程池里面的任务执行完毕并且主线程已经退出后，JVM仍然存在

<h3 id='14'>问题复现</h3>

```java
public class TestShutDown {

    static void asynExecuteOne(){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        executor.execute(new Runnable() {
            @Override
            public void run() {
                System.out.println("--async execute one");
            }
        });
    }

    static void asynExecuteTwo(){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        executor.execute(new Runnable() {
            @Override
            public void run() {
                System.out.println("--async execute two");
            }
        });
    }
      public static void main(String[]args){
        //(1)同步执行
        System.out.println("--sync execute");
        //(2)异步执行操作one
        asynExecuteOne();
        //(3)异步执行操作two
        asynExecuteTwo();
        System.out.println("---execute over");
    }

}

```

执行后发现ide内的小红块依旧存在，说明JVM没有退出，在terminal内执行ps -eaf|grep java命令发现Java进程还存在。加上 `executor.shutdown();`后发现JVM退出，这说明只有调用了线程池的shutdown方法后，线程池任务执行完毕，线程池的资源才会被释放。

<h3 id='15'>问题分析</h3>

基础篇hzong提到了守护线程与用户线程，JVM退出的条件后当前不存在用户线程，而线程池默认的ThreadFactory创建的线程是用户线程

```java
static calss DefaultThreadFactory implements ThreadFactory{
public Thread newThread(Runnable r) {
            Thread t =new Thread(group, r,namePrefix +threadNumber.getAndIncrement(),0);
            if(t.isDaemon()){
                t.setDaemon(false);
            }
            if(t.getPriority()!= Thread.NORM_PRIORITY){
                t.setPriority(Thread.NORM_PRIORITY);
            }
            return t;
        }
      }
```

由上面的代码可以知道，线程池默认的ThreadFacotry创建的都是用户线程，下面是shutdown的主要代码

```java
public void shutdown() {
        final ReentrantLock mainLock = this.mainLock;
        mainLock.lock();
        try {
            checkShutdownAccess();
            advanceRunState(SHUTDOWN);
            interruptIdleWorkers();
            onShutdown(); // hook for ScheduledThreadPoolExecutor
        } finally {
            mainLock.unlock();
        }
        tryTerminate();
    }
```

这里在shutdown方法里面设置了线程池的状态为SHUTDOWN，并且设置了所有worker空闲线程（阻塞到队列的take()方法的线程）的中断标志。那么下面来看在工作线程work而里面是不是设置了中断标志，然后它就会退出。

```java
final void runWoker(Worker w){
	...
    try{
      while(task!null||(task==getTask()!=null)){
        ...
      }
      ...
    }finally{
      ...
    }
}

private Runnable getTask(){
  boolean timeOut = false;
  for(; ;){
    ..
      (1)
      if(rs >= SHUTDOWN && (rs>=STOP)||workQueue.isEmpty()){
        decrementWorkerCount();
        return null;
      }
    	
    	try{
        //(2)
        Runnable r = timed?
          workQueue.poll(keepAliveTime,TimeUnit.NANOSECONDS);
        	workQueue.take();
        if(r!=null)return r;
        timeOut=true;
      }catch(InterruptedException retry){
        timeOut=false;
      }
    }
}
```

上述代码中，如果队列里没有任务，则工作线程被阻塞到代码（2）中等待从工作队列里面获取一个任务。这时候如果调用线程池的shutdown命令（shutdown命令会中断所有的工作线程），则代码（2）会抛出InterruptedException异常而返回，这个异常被捕捉到了，所以继续执行代码（1），而执行shutfown时设置了线程池的状态为SHUTDOWN，所以getTask方法返回了null，因而runWoker方法退出了循环，工作线程退出。

<h3 id='16'>小结</h3>

日常开发中使用线程池后一定要记得调用shutdown方法关闭

<h2 id='17'>
使用ThreadLocal不当可能会导致内存泄漏
</h2>

基础篇已经讲解了ThreadLocal的原理，本节着重介绍ThreadLocal会导致内存泄漏的原因，并给出使用ThreadLocal导致内存泄漏的案例。

<h3 id='18'>为何会出现内存泄漏</h3>

之前讲过，ThreadLocal只是一个工具类，具体存放变量的是线程的threadLocal变量。threadLocal是一个ThreadLocalMap类型的变量

![image-20191116143858190](/Users/hzx/Library/Application Support/typora-user-images/image-20191116143858190.png)

由上图可知，ThreadLocalMap内部是一个Entry数组，Entry继承自WeakReference，Entry内部的value用来存储通过ThreadLocal的set方法传递的值，那么ThreadLocal对象本身存放在哪里？看看Entry的构造函数

```java
Entry(ThreadLocal<?> k, Object v) {
                super(k);
                value = v;
            }
            
    public WeakReference(T referent) {
        super(referent);
    }
     Reference(T referent) {
        this(referent, null);
    }

 Reference(T referent, ReferenceQueue<? super T> queue) {
        this.referent = referent;
        this.queue = (queue == null) ? ReferenceQueue.NULL : queue;
    }
            
```

k被传递给WeakRerence的构造函数，也就说ThreadLocalMap里面的key为ThreadLocal对象的弱引用，具体就是referent变量引用了ThreadLocal对象，value为具体调用ThreadLocal的set方法时传递的值。

当一个线程调用ThreadLocal的set方法设置变量时，当前线程的TrheadLocalMap里就会存放一个记录，这个记录的key为ThreadLocal的弱引用，value则为其设置的值。如果当前线程一直存在且没有调用ThreadLocal的remove方法，并且这时候在其他地方还有对ThreadLocal的引用，则当前线程的ThreadLocalMap变量里面会存在对ThreadLocal变量的引用和对value对象的引用，它们是不会被释放的，这就造成了内存泄漏。

考虑这个 ThreadLocal变量没有其他强依赖，而当前线程还存在的情况， 由于线程的 ThreadLoca!Map 里面的 key 是弱依赖 ，所以当前线程的 ThreadLocalMap 里面的 ThreadLocal变量的弱引用会在 gc 的时候被回收，但是对应的 value还是会造成内存泄漏， 因为这时候 ThreadLoca!Map里面就会存在 key为 null但 value不为 null 的 entry项。 

其实在 ThreadLocal 的 set、 get和 remove方法里面可以找一些时机对这些 key为 null 的 entry进行清理 ， 但是这些清理不是必须发生的。下面简单说下 ThreadLocalMap 的 remove 方法中的清理过程。 

```java
private void remove(ThreadLocal<?> key) {
  				//(1)计算当前ThreadLocal变量所在的table数组位置，尝试使用快速定位方法
            Entry[] tab = table;
            int len = tab.length;
            int i = key.threadLocalHashCode & (len-1);
  			//(2)这里使用循环是防止快速定位失败后，遍历table数组
            for (Entry e = tab[i];
                 e != null;
                 e = tab[i = nextIndex(i, len)]) {
              //(3)找到
                if (e.get() == key) {
                  //(4)找到则调用WeakReference的clear方法清除对ThreadLocal的弱引用
                    e.clear();
                  //(5)清理key为null的元素
                    expungeStaleEntry(i);
                    return;
                }
            }
        }
```

代码（6）去掉了对value的引用，到这里当前线程里面的当前ThreadLocal对象的信息就被清理完了。

```java
private int expungeStaleEntry(int staleSlot) {
            Entry[] tab = table;
            int len = tab.length;

            //(6)expunge entry at staleSlot
            tab[staleSlot].value = null;
            tab[staleSlot] = null;
            size--;

            // Rehash until we encounter null
            Entry e;
            int i;
            for (i = nextIndex(staleSlot, len);
                 (e = tab[i]) != null;
                 i = nextIndex(i, len)) {
                ThreadLocal<?> k = e.get();
              //(7)
                if (k == null) {
                    e.value = null;
                    tab[i] = null;
                    size--;
                } else {
                    int h = k.threadLocalHashCode & (len - 1);
                    if (h != i) {
                        tab[i] = null;

                        // Unlike Knuth 6.4 Algorithm R, we must scan until
                        // null because multiple entries could have been stale.
                        while (tab[h] != null)
                            h = nextIndex(h, len);
                        tab[h] = e;
                    }
                }
            }
            return i;
        }
```

代码 (7) 从当前元素的下标开始查看 table数组里面是否有 key为 null的其他元素，有则清理。循环退出的条件是遇到 table里面有 null 的元素。 所以这里知道 null元素后面的 Entry里面 key为 null 的元素不会被清理。

总结 : ThreadLoca!Map的Entry 中的 key 使用的是对 ThreadLocal 对象的弱引用，这在避免内存泄漏方面是一个进步，因为如果是强引用， 即使其他地方没有对ThreadLocal 对象的引用， ThreadLocalMap 中的 ThreadLocal 对象还是不会被回收 ， 而如果是弱引用则 ThreadLocal 引用是会被回收掉的。但是对应的 value 还是不能被回收，这时候 ThreadLocalMap里面就会存在key为 null 但是 value不为null的entry 项 ， 虽然ThreadLocalMap提供了 set、get和remove方法，可以在一些时机下对这些Entry项进行清理，但是这是不及时的，也不是每次都会执行，所以在一些情况下还是会发生内存漏， 因此在使用完毕后及时调用 remove方法才是解决内存泄漏问题的王道。

<h3 id='19'>在线程池中使用ThreadLocal导致的内存泄漏</h3>

```java
public class MemoryLeaking {
    static class LocalVariable{
        private Long[]a = new Long[1024*1024];
    }

    //(1)
    final static ThreadPoolExecutor poolExecutor= new ThreadPoolExecutor(5,5,1, TimeUnit.MINUTES,new LinkedBlockingDeque<>());
    final static ThreadLocal<LocalVariable> localVariable = new ThreadLocal<LocalVariable>();

    public static void main (String[]args) throws  InterruptedException{
        //(3)
        for(int i=0;i<50;i++){
            poolExecutor.execute(new Runnable() {
                @Override
                public void run() {
                    //(4)
                    localVariable.set(new LocalVariable());
                    System.out.println("use local variable");
                    localVariable.remove();
                }
            });
            Thread.sleep(1000);

        }
        System.out.println("pool execute over");
    }

}

```

代码（1）创建了一个核心线程数和最大线程数都为5的线程池。代码（2）创建了一个ThreadLocal的变量，泛型参数为LocalVariable，LocalVariable内部是一个Long数组。代码（3）向线程池里面放入了50个任务。代码（4）设置当前线程的localVariable变量，也就是把new的LocalVariable变量放入当前线程的threadLocals变量中。

由于没有调用线程池的shutdown或者shutdownNow方法，所以线程池里面的用户线程不会退出，进而JVM也不会退出。

运行代码，使用jconsole监控内存变化，再去掉localVariable.remove()的注释，运行后观察堆内存变化。

由运行结果，没有调用remove方法主线程休眠时进程占用了大概70MB内存，调用后进程占用大概20MB内存，因此运行代码一时发生了内存泄漏

第一次运行代码时，在设置线程的 localVariable变量后没有调用 localVariable.remove()方法，这导致线程池里面 5 个核心线程的 threadLocals 变量里面的 new LocalVariable() 实例没有被释放。 虽然线程池里面的任务执行完了，但是线程池里面的 5 个线程会一直存在直到 JVM 进程被杀死。这里需要注意的是，由于 localVariable 被声明为了 static 变量，虽然在线程的 ThreadLocalMap里面对 localVariable进行了弱引用，但是 localVariable不会被回收。第二次运行代码时，由于线程在设置 localVariable变量后及时调用了 localVariable.remove()方法进行了清理，所以不会存在内存泄漏问题。

总结:如果在线程池里面设置了 ThreadLocal变量，则一定要记得及时清理，因为线程池里面的核心线程是一直存在的，如果不清理，线程池的核心线程的 threadLocals变量会一直持有 ThreadLocal 变量。